include/xmpm/discontinuity_3d.tcc

```diff
@@ -0,0 +1,169 @@
+bool mpm::Discontinuity3D<Tdim>::initialize(
+    const std::vector<VectorDim>& points,
+    const std::vector<std::vector<mpm::Index>>& surfs) {
+  bool status = true;
+  // Create points from file
+  bool point_status = this->create_points(points);
+  if (!point_status) {
+    status = false;
+    throw std::runtime_error(
+        "Addition of points in discontinuity to mesh failed");
+  }
+  // Create surfaces from file
+  bool surf_status = create_surfaces(surfs);
+  if (!surf_status) {
+    status = false;
+    throw std::runtime_error(
+        "Addition of surfaces in discontinuity to mesh failed");
+  }
+
+  bool normal_status = initialize_center_normal();
+  if (!normal_status) {
+    status = false;
+    throw std::runtime_error(
+        "initialization of the center and normal vector of the discontunity failed");
+  }
+
+  this->assign_point_friction_coef();
+  return status;
+};
+
+//! create surfaces from file
+template <unsigned Tdim>
+bool mpm::Discontinuity3D<Tdim>::create_surfaces(
+    const std::vector<std::vector<mpm::Index>>& surfs) {
+
+  bool status = true;
+  try {
+    // Check if surfs is empty
+    if (surfs.empty()) throw std::runtime_error("List of surfaces is empty");
+    // Iterate over all surfaces
+    for (const auto& points : surfs) {
+
+      mpm::discontinuity_surface<Tdim> surf(points);
+
+      surfaces_.emplace_back(surf);  //
+    }
+  } catch (std::exception& exception) {
+    console_->error("{} #{}: {}\n", __FILE__, __LINE__, exception.what());
+    status = false;
+  }
+  return status;
+}
+
+// initialize the center and normal of the surfaces
+template <>
+bool mpm::Discontinuity3D<3>::initialize_center_normal() {
+  bool status = true;
+  try {
+    VectorDim center;
+    VectorDim normal;
+    Eigen::Matrix<mpm::Index, 3, 1> points;
+
+    for (auto& surf: surfaces_) {
+      points = surf.points();
+
+      // the center of the surfaces
+      for (int i = 0; i < 3; i++)
+        center[i] = 1.0 / 3 *
+                    (points_[points[0]].coordinates()[i] +
+                     points_[points[1]].coordinates()[i] +
+                     points_[points[2]].coordinates()[i]);
+
+      surf.assign_center(center);
+
+      // the normal of the surfaces
+      normal = three_cross_product(points_[points[0]].coordinates(),
+                                   points_[points[1]].coordinates(),
+                                   points_[points[2]].coordinates());
+      double det = std::sqrt(normal[0] * normal[0] + normal[1] * normal[1] +
+                             normal[2] * normal[2]);
+      normal = 1.0 / det * normal;
+
+      surf.assign_normal(normal);
+    }
+  } catch (std::exception& exception) {
+    console_->error("{} #{}: {}\n", __FILE__, __LINE__, exception.what());
+    status = false;
+  }
+  return status;
+}
+
+// return the cross product of ab X bc
+template <unsigned Tdim>
+Eigen::Matrix<double, Tdim, 1> mpm::Discontinuity3D<Tdim>::three_cross_product(
```

Comment:
Should this function be in `utilities` as well? @kks32 , what do you think?

Reply:
I agree with @thiagordonho, this can be put outside as generic function